File: 2025/src/day02/day02.exs
defmodule Day02 do
  @path "src/day02/input.txt"

  def main do
    raw_file = File.read!(@path)
    ranges = parse_ranges(raw_file)
    all_ids = individual_numbers(ranges)
    part1(all_ids) |> IO.inspect(label: "Part 1: strings that are mirrored")
    part2(all_ids) |> IO.inspect(label: "Part 2: strings that ahve repeating patterns")
  end

  def part1(ranges) do
    ranges
    |> Enum.filter(&int_doubled?/1)
    |> Enum.sum()
  end

  def part2(ranges) do
    ranges
    |> Enum.filter(&int_repeats?/1)
    |> Enum.sum()
  end

  def parse_ranges(string) do
    string
    |> String.split(",", trim: true)
    |> Enum.map(fn range ->
      [start_str, end_str] = String.split(range, "-", trim: true)
      {String.to_integer(start_str), String.to_integer(end_str)}
    end)
  end

  def individual_numbers(ranges) do
    Enum.flat_map(ranges, fn {start_range, end_range} ->
      start_range..end_range
    end)
  end

  defp int_doubled?(id) when is_integer(id), do: id |> Integer.digits() |> doubled?()

  defp doubled?(digits) do
    {first_half, second_half} = Enum.split(digits, div(length(digits), 2))
    first_half == second_half
  end

  defp int_repeats?(id) when is_integer(id), do: id |> Integer.digits() |> repeats?()

  defp repeats?(digits) when length(digits) >= 2 do
    len = length(digits)
    max_pattern_len = div(len, 2)

    Enum.any?(1..max_pattern_len, fn pattern_len ->
      pattern = Enum.take(digits, pattern_len)
      is_pattern_repeated?(digits, pattern)
    end)
  end

  defp repeats?(_), do: false

  defp is_pattern_repeated?(digits, pattern) do
    digits
    |> Enum.chunk_every(length(pattern))
    |> Enum.all?(&(pattern == &1))
  end

  ### Originally I exhaustively matched all cases for the max length
  # defp doubled?([a, a]), do: true
  # defp doubled?([a, b, a, b]), do: true
  # defp doubled?([a, b, c, a, b, c]), do: true
  # defp doubled?([a, b, c, d, a, b, c, d]), do: true
  # defp doubled?([a, b, c, d, e, a, b, c, d, e]), do: true
  # defp doubled?([a, b, c, d, e, f, a, b, c, d, e, f]), do: true
  # defp doubled?(_), do: false
  # defp repeats?([a, a]), do: true
  # defp repeats?([a, a, a]), do: true
  # defp repeats?([a, a, a, a]), do: true
  # defp repeats?([a, a, a, a, a]), do: true
  # defp repeats?([a, a, a, a, a, a]), do: true
  # defp repeats?([a, a, a, a, a, a, a]), do: true
  # defp repeats?([a, a, a, a, a, a, a, a]), do: true
  # defp repeats?([a, a, a, a, a, a, a, a, a]), do: true
  # defp repeats?([a, a, a, a, a, a, a, a, a, a]), do: true
  # defp repeats?([a, b, a, b]), do: true
  # defp repeats?([a, b, a, b, a, b]), do: true
  # defp repeats?([a, b, a, b, a, b, a, b]), do: true
  # defp repeats?([a, b, a, b, a, b, a, b, a, b]), do: true
  # defp repeats?([a, b, a, b, a, b, a, b, a, b, a, b]), do: true
  # defp repeats?([a, a, b, b, a, a, b, b]), do: true
  # defp repeats?([a, a, b, b, a, a, b, b, a, a, b, b]), do: true
  # defp repeats?([a, a, a, a, b, b, b, b]), do: true
  # defp repeats?([a, a, a, a, b, b, b, b, a, a, a, b, b, b]), do: true
  # defp repeats?([a, b, c, a, b, c]), do: true
  # defp repeats?([a, b, c, a, b, c, a, b, c]), do: true
  # defp repeats?([a, b, c, a, b, c, a, b, c, a, b, c]), do: true
  # defp repeats?([a, a, b, b, c, c, a, a, b, b, c, c]), do: true
  # defp repeats?([a, a, b, b, c, c, a, a, b, b, c, c, a, a, b, b, c, c]), do: true
  # defp repeats?([a, b, c, d, a, b, c, d]), do: true
  # defp repeats?([a, b, c, d, a, b, c, d, a, b, c, d]), do: true
  # defp repeats?([a, b, c, d, a, b, c, d, a, b, c, d, a, b, c, d]), do: true
  # defp repeats?([a, a, b, b, c, c, d, d, a, a, b, b, c, c, d, d]), do: true
  # defp repeats?([a, b, c, d, e, a, b, c, d, e]), do: true
  # defp repeats?([a, b, c, d, e, a, b, c, d, e, a, b, c, d, e]), do: true
  # defp repeats?([a, b, c, d, e, f, a, b, c, d, e, f]), do: true
  # defp repeats?([a, b, c, d, e, f, a, b, c, d, e, f, a, b, c, d, e, f]), do: true
  # defp repeats?(_), do: false
end

Day02.main()
ExUnit.start()

defmodule Day02Test do
  use ExUnit.Case, async: true

  @example_input "11-22,95-115,998-1012,1188511880-1188511890,222220-222224,1698522-1698528,446443-446449,38593856-38593862,565653-565659,824824821-824824827,2121212118-2121212124"

  test "part1" do
    result = @example_input |> Day02.parse_ranges() |> Day02.individual_numbers() |> Day02.part1()
    assert result == 1_227_775_554
  end

  test "part2" do
    result = @example_input |> Day02.parse_ranges() |> Day02.individual_numbers() |> Day02.part2()
    assert result == 4_174_379_265
  end
end

File: 2025/src/day02/day02.rs
use std::fs;

const INPUT_PATH: &str = "src/day02/input.txt";

fn main() -> (u64, u64) {
    let raw_file = fs::read_to_string(INPUT_PATH).expect("Failed to read input file");
    let ranges = parse_ranges(&raw_file);
    let all_ids = individual_numbers(ranges);

    let part1 = part1(&all_ids);
    println!("Part 1: strings that are mirrored: {}", part1);
    let part2 = part2(&all_ids);
    println!("Part 2: strings that have repeating patterns: {}", part2);

    (part1, part2)
}

fn part1(ids: &[u64]) -> u64 {
    ids.iter().filter(|&&id| int_doubled(id)).sum()
}

fn part2(ids: &[u64]) -> u64 {
    ids.iter().filter(|&&id| int_repeats(id)).sum()
}

fn parse_ranges(raw_file: &str) -> Vec<(u64, u64)> {
    raw_file
        .split(",")
        .map(|range| {
            let (start, stop) = range.split_once("-").unwrap();
            (start.parse::<u64>().unwrap(), stop.parse::<u64>().unwrap())
        })
        .collect()
}

fn individual_numbers(ranges: Vec<(u64, u64)>) -> Vec<u64> {
    ranges
        .into_iter()
        .flat_map(|(start, stop)| start..=stop)
        .collect()
}

fn int_doubled(id: u64) -> bool {
    let digits = id.to_string();

    doubled(digits)
}

fn doubled(digits: String) -> bool {
    let (first_half, second_half) = digits.split_at(digits.len() / 2);
    first_half == second_half
}

fn int_repeats(id: u64) -> bool {
    let digits = id.to_string().chars().collect();

    repeated(digits)
}

fn repeated(digits: Vec<char>) -> bool {
    let len = digits.len();
    if len < 2 {
        return false;
    }

    let max_pattern_len = len / 2;

    (1..=max_pattern_len).any(|len| {
        let pattern = digits.clone().into_iter().take(len).collect();
        is_repeated_pattern(digits.clone(), pattern)
    })
}

fn is_repeated_pattern(digits: Vec<char>, pattern: Vec<char>) -> bool {
    digits.chunks(pattern.len()).all(|p| p == pattern)
}

#[cfg(test)]
mod tests {
    use super::*;

    const EXAMPLE_INPUT: &str = "11-22,95-115,998-1012,1188511880-1188511890,222220-222224,1698522-1698528,446443-446449,38593856-38593862,565653-565659,824824821-824824827,2121212118-2121212124";

    #[test]
    fn test_day02_main() {
        let (part1, part2) = main();
        assert_eq!(part1, 28146997880);
        assert_eq!(part2, 40028128307);
    }

    #[test]
    fn test_day02_part1() {
        let ranges = parse_ranges(EXAMPLE_INPUT);
        let all_ids = individual_numbers(ranges);
        let result = part1(&all_ids);
        assert_eq!(result, 1227775554);
    }

    #[test]
    fn test_day02_part2() {
        let ranges = parse_ranges(EXAMPLE_INPUT);
        let all_ids = individual_numbers(ranges);
        let result = part2(&all_ids);
        assert_eq!(result, 4174379265);
    }
}
